src/System.Text.Primitives/System/Text/Encoders/Utf8Util.Validation.cs

```diff
@@ -0,0 +1,728 @@
+                    // ED9F = 1110 1101 1001 1111 => invalid (surrogate) patterns are 1110 1101 101# ####
+                    // If using the bitmask ......................................... 0000 1111 0010 0000 (=0F20),
+                    // Then invalid (overlong) patterns match the comparand ......... 0000 0000 0000 0000 (=0000),
+                    // And invalid (surrogate) patterns match the comparand ......... 0000 1101 0010 0000 (=0D20).
+
+                    if (BitConverter.IsLittleEndian)
+                    {
+                        // The "overlong or surrogate" check can be implemented using a single jump, but there's
+                        // some overhead to moving the bits into the correct locations in order to perform the
+                        // correct comparison, and in practice the processor's branch prediction capability is
+                        // good enough that we shouldn't bother. So we'll use two jumps instead.
+
+                        // Can't extract this check into its own helper method because JITter produces suboptimal
+                        // assembly, even with aggressive inlining.
+
+                        uint comparand = thisDWord & 0x0000200FU;
+                        if ((comparand == 0U) || (comparand == 0x0000200DU)) { goto Error; }
+                    }
+                    else
+                    {
+                        uint comparand = thisDWord & 0x0F200000U;
+                        if ((comparand == 0U) || (comparand == 0x0D200000U)) { goto Error; }
+                    }
+
+                    ProcessSingleThreeByteSequenceSkipOverlongAndSurrogateChecks:
+
+                    inputBufferCurrentOffset += 3;
+                    inputBufferRemainingBytes -= 3;
+                    tempRuneCount -= 2; // 3 bytes -> 1 rune
+
+                    // Occasionally one-off ASCII characters like spaces, periods, or newlines will make their way
+                    // in to the text. If this happens strip it off now before seeing if the next character
+                    // consists of three code units.
+
+                    if (DWordFourthByteIsAscii(thisDWord))
+                    {
+                        inputBufferCurrentOffset += 1;
+                        inputBufferRemainingBytes--;
+                    }
+
+                    SuccessfullyProcessedThreeByteSequence:
+
+                    // Optimization: A three-byte character could indicate CJK text, which makes it likely
+                    // that the character following this one is also CJK. We'll try to process several
+                    // three-byte sequences at a time.
+
+                    if (IntPtr.Size >= 8 && BitConverter.IsLittleEndian && inputBufferRemainingBytes >= (sizeof(ulong) + 1))
+                    {
+                        ulong thisQWord = Unsafe.ReadUnaligned<ulong>(ref Unsafe.Add(ref inputBuffer, inputBufferCurrentOffset));
```

Comment:
`thisULong` instead of `thisQWord`? Similar for other occurrences.

Reply:
Or rather this UInt64